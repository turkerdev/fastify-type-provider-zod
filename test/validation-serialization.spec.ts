/**
 * @jest-environment node
 */

import { describe, expect, it } from '@jest/globals';
import type { FastifyInstance } from 'fastify';

import makeServer from './core/server';

let server: FastifyInstance;

describe('Fastify Type Provider Zod', () => {
  describe('Validation', () => {
    it('should be able to tell that validation is required', async () => {
      server = await makeServer();

      const response = await server.inject({
        method: 'GET',
        url: '/test-valid-serialization',
      });

      expect(response.statusCode).toBe(400);
      expect(JSON.parse(response.body)).toMatchSnapshot();
    });

    it('should be able to tell that validation need to have a number', async () => {
      server = await makeServer();

      const response = await server.inject({
        method: 'GET',
        url: '/test-valid-number?value=foo',
      });
      expect(response.statusCode).toBe(400);
      expect(JSON.parse(response.body)).toMatchSnapshot();
    });
  });

  describe('Serialization', () => {
    it('should return the result', async () => {
      server = await makeServer();

      const response = await server.inject({
        method: 'GET',
        url: '/test-valid-serialization?value=test',
      });

      expect(response.statusCode).toBe(200);
    });

    it('should tell that response is not valid', async () => {
      server = await makeServer();

      const response = await server.inject({
        method: 'GET',
        url: '/test-error-serialization?value=test',
      });
      expect(response.statusCode).toBe(500);
      expect(JSON.parse(response.body)).toMatchObject({
        error: 'Internal Server Error',
        message: "Response doesn't match the schema",
        statusCode: 500,
      });
    });
  });

  describe('Mercurius', () => {
    it('should be able to consult /graphql', async () => {
      server = await makeServer();

      const response = await server.inject({
        method: 'GET',
        url: '/graphql',
      });

      expect(response.statusCode).toBe(400);
      expect(JSON.parse(response.body)).toMatchObject({
        data: null,
        errors: [
          {
            message: 'Unknown query',
          },
        ],
      });
    });
  });

  describe('@fastify/swagger', () => {
    it('should be able to return a json', async () => {
      server = await makeServer();

      const response = await server.inject({
        method: 'GET',
        url: '/documentation/json',
      });
      expect(response.statusCode).toBe(200);
    });
  });
});
