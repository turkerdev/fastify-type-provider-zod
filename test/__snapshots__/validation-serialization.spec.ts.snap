// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fastify Type Provider Zod Validation should be able to tell that validation is required 1`] = `
{
  "code": "FST_ERR_VALIDATION",
  "error": "Bad Request",
  "message": "[
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "value"
    ],
    "message": "Required"
  }
]",
  "statusCode": 400,
}
`;

exports[`Fastify Type Provider Zod Validation should be able to tell that validation need to have a number 1`] = `
{
  "code": "FST_ERR_VALIDATION",
  "error": "Bad Request",
  "message": "[
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [
      "value"
    ],
    "message": "Expected number, received string"
  }
]",
  "statusCode": 400,
}
`;
